### ISSUE TITLE
 docs need a complete rework

### ISSUE THREAD

**Issue by yelouafi** [yelouafi/redux-saga#70]
Actual documentation was mainly intended to showcase the library capabilities and how it is different from other middlewares. I think docs need to be reworked with a more 'conventional'  approach

My initial list

1- API reference
2- Step by step tutorials
3- Section on unit testing Sagas (maybe also integration testing with store/middleware)
4- Examples on how to manage common concurrency patterns (watch and fork, watch and fork latest...)
5- Example solutions for some common flows (authorization, infinite scroll ...)
6- Links to external resources (good tutorials on Generators, redux-saga articles, app examples)

Docs will use Gitbook format

Any further ideas ?

**Comment by slorber** [comment]
agreed.

Also explaining why the term "saga" was initially choosen will help people understand how it relates to existing backend concepts

**Comment by youknowriad** [comment]
May be you should consider https://tonicdev.com for code snippets. I have to admit I've never used it yet, but I was really impressed by the possibilities (embed a live running code).

It have also a Gitbook plugin.

**Comment by paul42** [comment]
I'd like to :+1: the 5 - example solutions for some common flows - I'm still just digging in and haven't done much more than a hello world for react and redux, so if it's not too much trouble the examples would help me the best. maybe even a 'hello world' example for people like myself who like to 'code along' and see it work.

**Comment by vire** [comment]
Do you think a transition guide from `redux-thunk` would be useful? Because I've started with thunks but they're hard to maintain when the app grows... (+hard to test)

**Comment by cherta** [comment]
I like the idea! I think people learn by example, so if we can put together some cook book it will be awesome, I guess it similar to what @yelouafi says on point 5.

Just for future reference I posted [an article](https://medium.com/@iamcherta/redux-sagas-and-urls-40daeec8eb0b) of how I solved a particular issue with router, I guess is mostly the same I mentioned on #37.

Count on me to create some of this docs.

**Comment by slorber** [comment]
@yelouafi I was thinking or organising some kind of "challenge" like the one of @staltz: https://github.com/staltz/flux-challenge

The challenge idea was to plug an onboarding to an existing TodoMVC application. 
The onboarding should be complex enough to showcase the power of different approaches, and I think it would be a nice tutorial to display redux-saga capabilities.
Just haven't figured out yet the onboarding spec to implement :)

**Comment by cef62** [comment]
I like the idea of better explain the term _saga_ proposed by @slorber. A first good step could be to split the current readme in smaller docs and organize them. 

If you like the idea of using gitbook I can prepare an initial setup to use it. I like the approach of redux where docs is available both as plain markdown files and as compiled gitbook.

I could work on a PR tomorrow night.

**Comment by yelouafi** [comment]
@slorber Yes I think also Sagas should be introduced.

@cef62 I agree. will be a good start. Thanks for working on this

@cherta Thanks. Integration with the router would make a good example.

**Comment by yelouafi** [comment]
>Do you think a transition guide from redux-thunk would be useful?

Yes perhaps the tutorial should adopt this approach. to make learning as smooth as possible

**Comment by nirrek** [comment]
> 5- Example solutions for some common flows (authorization, infinite scroll ...)

In this section I think it would be useful to show how to adapt event-based APIs for consumption inside sagas. Websockets, Firebase, Web Audio/Video etc. only expose EventEmitter-style APIs. Wrapping these to provide a pull-based interface that can be used in sagas will be a fairly common pattern.

**Comment by yelouafi** [comment]
FYI, thanks to @cef62 actual docs were migrated to GitBook. docs are now on this link

http://yelouafi.github.io/redux-saga/

Any suggestions for the TOC are welcome

Also added a section for API reference

http://yelouafi.github.io/redux-saga/docs/api/index.html

Still need some improvements but better than nothing

**Comment by gaearon** [comment]
Amazing work :+1:

**Comment by yelouafi** [comment]
@gaearon thanks! Hope to add more content next week

**Comment by gaearon** [comment]
I think it's already a good idea to put the Gitbook into the "website" field at the top of the repo to give it visibility.

**Comment by yelouafi** [comment]
@gaearon done. thanks

**Comment by staltz** [comment]
I came here because of https://twitter.com/dan_abramov/status/696689725797769216 and would like to just point out one thing:

> An alternative Side Effect model for Redux applications. Instead of dispatching thunks which get handled by the redux-thunk middleware, you create Sagas to gather all your Side Effects logic in a central place.
>This means the logic of the application lives in 2 places:
> - Reducers are responsible for handling state transitions between actions
> - Sagas are responsible for orchestrating complex/asynchronous operations.
> Sagas are created using Generator functions.

Avoid heavy beginner-unfriendly jargon for the first text they will ever read. Drop "saga" (why not rename it to "Redux coprocesses"?), drop "side effect model", drop "dispatching thunks", drop "redux-thunk middleware".

Here's a suggestion:

"*Redux Coprocesses is a library for listening to actions, spawning background work, and dispatching results, ideal for HTTP requests and other side effects*"

Trust me, I have experience explaining Cycle.js in a weird way ("Honestly reactive framework for Human-Computer Interaction"), I learned how to keep it dumb ("Functional reactive JavaScript framework for cleaner code").

Just my cents, make your own mind about it :+1:

**Comment by vire** [comment]
@staltz I like your suggestion. Despite the fact I like *saga* as something special, your point is valid and it's acceptable for wider audience.

**Comment by staltz** [comment]
This is what my brain understands when I parse the word "saga":

![casjfhvweaa-pp3](https://cloud.githubusercontent.com/assets/90512/12887317/22a18426-ce7c-11e5-9505-16402b9b9af2.png)

http://www.thefreedictionary.com/saga

:)

**Comment by slorber** [comment]
@staltz the name "saga" does not come from nowhere, it actually already is used in the backend event-sourcing and domain-driven-design communities (also called "process manager"). It is used to orchestrate complex operations (sometimes eventually consistent transactions) across differend decoupled bounded contexts. 

A bounded context has generally no coupling to other bounded contexts and for the backend it is generally a good candidate to be isolated as a microservice. I think the closest thing we have in Redux community, is a "widget" or a "[ducks](https://github.com/erikras/ducks-modular-redux)". 

Besides being used for ajax interactions, the saga is also a mean to couple 2 ducks together in a meaningful way for your business/app, without creating cyclic dependencies between the 2 ducks.


So yes it takes some time to grasp and there is jargon to learn. Notice that even for a domain-driven-design newcomer, this jargon is quite hard to follow at first :) 
I agree that it should be made simpler for newcomers. I've read many times in the DDD community that people sum it up this way: 

> A saga is just a piece of software that receive events and may issue new commands

But the UI is different from the backend world (no real need for commands, and a lot more work with remote services, ie ajax)

For FP users we could also point out that the implementation shares many similarities with the [Free monad](http://underscore.io/blog/posts/2015/04/14/free-monads-are-simple.html). Different user background -> different way to sell the lib. For my background "free monad" is more meaningful than "coprocess"

---------------------------------------------------------------------

I agree for a better "tagline" and also think that showing on the first page some very common and easy to understand "recipes" would be nice  (the original async counter example does not really speak to me)

However I still think that for more advanced usecase this jargon can be useful to dig deeper into this subject. Flux/Redux is really close to event-sourcing, and as time pass we see more and more existing backend patterns being used in the frontend (including this project). Maybe it's time to simply recognize this fact and stop duplicating already existing jargon?

**Comment by yelouafi** [comment]
@staltz Thanks for you feedback. I agree that the Intro should adopt more familiar terms like Asynchronous actions, workflow, background tasks ...

For the Saga term, it's true that Sagas as used in the backend seem different from the frontend (long running business transactions, failure management ...) and may put people in confusion. But dropping it completely seems a bit radical. Even in the backend people appear to use Sagas in different contexts (I think the same thing goes for FRP if we take into consideration how the concept has evolved since the original conal elliott's paper).

It seems to me that the steep learning curve has more to do with the amount of new things a new user has to digest (Generators, async Generators, event pulling, declarative Effects, concurrency management ...). And I think the focus should be on explaining those concepts and putting them in context (tutorials, common cases examples)

**Comment by yelouafi** [comment]
@slorber 
>the original async counter example does not really speak to me

totally agree; and it's broken en plus.

**Comment by gaearon** [comment]
>Trust me, I have experience explaining Cycle.js in a weird way ("Honestly reactive framework for Human-Computer Interaction"), I learned how to keep it dumb ("Functional reactive JavaScript framework for cleaner code").

:+1: 

Also, Redux used to be “Atomic Flux with hot reloading”. :open_mouth: 
So glad we changed that.

**Comment by gaearon** [comment]
As for the examples, here’s a few that would make sense to me:

* Closing modal windows on backdrop click -or- request to open another modal window
* Reauthenticating the user when the token is expired
* Showing the onboarding tooltips
* Showing a “delete item” confirmation dialog and dispatching an action on confirmation
* Fetching new data in response to the route changes

**Comment by gaearon** [comment]
>drop "dispatching thunks", drop "redux-thunk middleware".

I think we can drop those when they are no longer dominant way of performing fetches, etc, in Redux. For now we need to make it clear that Redux Saga handles those scenarios and obsoletes Redux Thunk which is pretty much the most popular Redux middleware at this time.

>Avoid heavy beginner-unfriendly jargon for the first text they will ever read. Drop "saga" (why not rename it to "Redux coprocesses"?), drop "side effect model"

Is “saga” more intimidating than “coprocess”? I have barely any exposure to either, and both seem equally vague to me. I like that “sagas” haves precedent in the event sourcing. Since I’m both to new terms, “saga” sounds even a bit friendlier to me than “coprocess” or “daemon”. It’s easier to write :smile: .

I think that we need to pick a few terms we consider essential for the library. For example, **saga**, **effect**, and **task** seem like good candidates. The library can’t really be explained in a good way if one refuses to acknowledge that these types of things exist and make specific sense in the context of the library (and correspond to types). However, once we determine those terms, we need to be careful to only mention these types in the documentation, and not anything else.

**Comment by yelouafi** [comment]
>The library can’t really be explained in a good way if one refuses to acknowledge that these types of things exist and make specific sense in the context of the library (and correspond to types). However, once we determine those terms, we need to be careful to only mention these types in the documentation, and not anything else.

That what I think. The goal is not hide 'difficult concepts' from the developer, Instead we should introduce them in a way that's both progressive and accessible.

The main difficulty I'm encountering now (to write a tutorial) is how to introduce things one by one, even a simple Counter example needs multiple concepts at once (Declarative Effects, the event pull concept, forking if we need to provide a correct example)

**Comment by cherta** [comment]
> The main difficulty I'm encountering now (to write a tutorial) is how to introduce things one by one, even a simple Counter example needs multiple concepts at once (Declarative Effects, the event pull concept, forking if we need to provide a correct example)

Is there a need to explain things one by one? I think is a better approach to show a problem explain one thing then upgrade the example. I guess a tutorial can push explaining declarative effects until it shows how to write test, on that scenario declarative effects make sense and it is more likely a person understands the concept that way.

**Comment by aams** [comment]
AS James D. Cook says: _"You can’t avoid state, but you can partition the stateful and stateless parts of your code. 100% functional purity is impossible, but 85% functional purity may be very productive."_

We eliminate "side effects" when coding pure functions so our code is easier to understand and test. Where our functions meet the real world, we are dealing with "effects", not "side effects".

But "effects" seems such a weak term to describe the interface to the environment in which an application runs. A saga is a story, so maybe "effects" are "story lines" or sub-plots, definitely not "side effects".

**Comment by johnsoftek** [comment]
AS James D. Cook says: _"You can’t avoid state, but you can partition the stateful and stateless parts of your code. 100% functional purity is impossible, but 85% functional purity may be very productive."_

We eliminate "side effects" when coding pure functions so our code is easier to understand and test. Where our functions meet the real world, we are dealing with "effects", not "side effects".

But "effects" seems such a weak term to describe the interface to the environment in which an application runs. A saga is a story, so maybe "effects" are "story lines" or sub-plots, definitely not "side effects".

**Comment by yelouafi** [comment]
Suggestions for a new Introductions ? 

Should at least keep the terms mentioned by @gaearon (Saga, Effect, Background tasks). 

Need to explain what Sagas are about in familiar terms. For now I'm thinking of

- Simplify Async control flow
- Easy testing
- Decouple Components from Reducers

**Comment by gaearon** [comment]
Something that came up in chat with @ForbesLindsay that I believe is important.

https://twitter.com/forbeslindesay/status/697428202894053376

Are we sure we are explaining concurrent fetches soon enough? Are the early examples misleading because concurrent actions are ignored but people are not used to this model?

**Comment by yelouafi** [comment]
@gaearon I'm aware of this, I discussed this with @ForbesLindesay in #68. That's what I meant by

>even a simple Counter example needs multiple concepts at once (Declarative Effects, the event pull concept, forking if we need to provide a correct example)

I think he's right about the examples being misleading even If they're short snippets. I'm currently writing an introductory tutorial and managing concurrency is planned here by pointing out the issue with blocking calls.

I think concurrency should be introduced early, and for the examples which makes blocking calls (like in `call` docs) put a warning about it.

Another complementary solution is by providing higher level abstractions that are *safe* to use for common cases (watch, watchLatest and maybe more see https://github.com/yelouafi/redux-saga/issues/68#issuecomment-178086859). I'm thinking of adding a redux-saga-contrib repo which provides those helpers and use them in the examples and repo demos when possible.

Enforcing the non blocking calls in the core library is complicated. as @slorber commented in #68, the library provides low level functions for concurrency which make it possible to implement complicated flows which are not trivial to write using simple watch or watchLatest

My point is that *we should not hide* race concerns from developers; instead make the library users aware of them in the best way. thinking about race conditions is inavoidable even in the most obvious cases. A simple example is concurrent requests using redux-thunk: the default behavior is not 100% safe because it doesn't handle out of order responses (ie application state could be updated with outdated response).

**Comment by slorber** [comment]
@yelouafi maybe watch/watchLatest should be in core and be one of the first example.

This will be used extensively by people wanting to replace ajax calls in actionCreators thunks, and most people think redux-saga is essentially to do only that, in a more testable way.

Idiomatic ajax thunk replacement is watch, but the doc should also point out that watchLatest can be better in certain usecases (ie not the counter example) and explain the differences, maybe by comparing the 2 low-level implementations so that people finally understand the concurrency concerns that @ForbesLindesay pointed out

**Comment by yelouafi** [comment]
>maybe watch/watchLatest should be in core and be one of the first example.

In thinking about it, perhaps this is the most suitable solution.

**Comment by gaearon** [comment]
I think splitting this into two repos is not a very good solution.
Figure out which utilities are needed in 95% of apps and include them in the core.

**Comment by yelouafi** [comment]
>Figure out which utilities are needed in 95% of apps and include them in the core.

For now it seems the most common needs are watch (which provides thunk like behavior) and watchLatest(which handles out of order responses by taking only the last response).

Probably another utility is one that outputs the responses in the same order of the requests

Rx and Bacon.js can also serve as a good reference (the flatMap and family functions)

**Comment by yelouafi** [comment]
>and be one of the first example.

Do you think the introductory tutorial should only use the Higher level API ?

**Comment by slorber** [comment]
> watchLatest(which handles out of order responses by taking only the last response).

You mean it only takes the response of the last query I guess?

> Do you think the introductory tutorial should only use the Higher level API ?

probably not but the reader should probably learn very soon about watch/watchLatest as it cover an usercase everybody has.

I think it would be nice to have a 2 column table with these 2 higher level function and the corresponding implementations details + how to test them, so that people understand well the distinction and how it got implemented

**Comment by gaearon** [comment]
I think `watch` examples should be on the front page. That’s what people come from. We can introduce generators later.

>I think it would be nice to have a 2 column table with these 2 higher level function and the corresponding implementations details + how to test them, so that people understand well the distinction and how it got implemented

Agreed.

**Comment by yelouafi** [comment]
>You mean it only takes the response of the last query I guess?

Yes

>I think watch examples should be on the front page. That’s what people come for. We can introduce generators later.

But still, the developer needs to implement the watch logic using Generators

**Comment by slorber** [comment]
```javascript
function* fetchUser(action) {
   try {
      const user = yield call(Api.fetchUser,action.payload.userId);
      yield put({type: "USER_FETCH_SUCCEEDED", user: user});
   } catch (e) {
      yield put({type: "USER_FETCH_FAILED",message: e.message});
   }
}

// Allow concurrent fetches of user if multiple "USER_FETCH_REQUESTED" gets dispatched very close to each others
watch("USER_FETCH_REQUESTED",fetchUser);

// Do not allow concurrent fetches of user
// If "USER_FETCH_REQUESTED" gets dispatched while a fetch is already pending, that pending fetch is cancelled
watchLatest("USER_FETCH_REQUESTED",fetchUser);
```

The watch logic can be a simple Ajax request for the first example as it is simple and match a very common usecase too.
Just with that snippet, users will know how to effectively replace their thunks and have better concurrency control.

**Comment by yelouafi** [comment]
Do you think `watch` and  `watchLatest` should return something ?

The denotation seems like a dynamic array of tasks. Should it be something with `getTasks()` method (in case of `watchLatest` it's only one task: the latest)  ?

**Comment by slorber** [comment]
Not necessarily, this is just pseudo code but maybe.

**Comment by yelouafi** [comment]
An idea is to make the middleware accept also watch Effects besides Saga

```javascript
// Allow concurrent fetches of user 
export const watchUserRequests = watch("USER_FETCH_REQUESTED",fetchUser);

// Do not allow concurrent fetches of user
// If "USER_FETCH_REQUESTED" gets dispatched 
// while a fetch is already pending, that pending fetch is cancelled 
// and only the latest one will be run
export const watchUserRequests = yield watchLatest("USER_FETCH_REQUESTED",fetchUser);

// in another module
import { watchUserRequests, someSaga } from './sagas'

const sagaMiddleware = createSagaMiddleware(watchUserRequests, someSaga)
```

**Comment by yelouafi** [comment]
Question about the naming, what would be more suitable

```javascript
watch(REQUEST, fetch)
watchLatest(REQUEST, fetch)
```
vs
```javascript
on(REQUEST, fetch)
onLatest(REQUEST, fetch)
```

**Comment by gaearon** [comment]
It’s important the APIs are googleable and don’t grab very generic names. It would also be great if there was an easy way to tell between effects (`take`, `put`) and saga helpers. Finally, I’m not completely for anything that hides that top-level sagas are generators because this will make people assume `watch` is the only way to create them. That’s the reason we didn’t ship popular `createReducer()` helper in Redux.

**Comment by yelouafi** [comment]
>It would also be great if there was an easy way to tell between effects (take, put) and saga helpers.

My initial idea was to make them higher order functions which return Sagas. But it has a minor side effect, the returned sagas will be anonymous Generator functions so it may make debugging and monitoring less obvious.

>Finally, I’m not completely for anything that hides that top-level sagas are generators because this will make people assume watch is the only way to create them. 

You're right. Finally I don't feel that solution either

**Comment by gaearon** [comment]
Just throwing this out there.
Not sure if I made mistakes in generator usage.

```js
// Separate effects and generators
import { takeEvery, takeLatest } from 'redux-saga'
import { call, put, take, getState } from 'redux-saga/effects'

function* logActions() {
  while (true) {
    const action = yield take(action);
    // getState() is an effect so it doesn’t need to be passed down
    const state = yield getState();
    console.log('action: ', action);
    console.log('state after: ', state);
  }
}

function* fetchUser(action) {
   try {
      const user = yield call(Api.fetchUser,action.payload.userId);
      yield put({type: "USER_FETCH_SUCCEEDED", user: user});
   } catch (e) {
      yield put({type: "USER_FETCH_FAILED",message: e.message});
   }
}

// takeEvery() returns a generator so we have a nice named function here.
function* fetchUsers() {
  // No need to support a special watch effect: use yield* to delegate.
  yield* takeEvery('USER_FETCH_REQUESTED', fetchUser);
}

// Export one root saga rather than an array.
// We already allow yield [] so it matches how you’d export one root reducer.
// Since getState() is an effect, we no longer need to thread it through sagas.
export default function* saga() {
  yield [
    logActions(),
    fetchUsers()
  ]
}
```

**Comment by yelouafi** [comment]
>``` yield* takeEvery('USER_FETCH_REQUESTED', fetchUser);```

Generator delegation solves the issue with anonymous functions. And as you said we don't need to implement a watch effect.

this also would make the implementation much easier

>`yield getState();`

this solve the issue partly (In the real-world demo for example, the root Saga passes different selectors to sub-sagas). But I guess still better than passing getState around

I think @slorber could have some objections here :)

For the last example, the root Saga won't terminate until all called Sagas terminate. It could be desirable but wanted just to mention that `saga` will also terminate if an Error bubbles from one of the child sagas

**Comment by yelouafi** [comment]
BTW maybe a naive question; how can I make this possible

```javascript
import ... from 'redux-saga/effects'
```

Until now I can only achieve this

```javascript
import ... from 'redux-saga/lib/effects'
```

**Comment by gaearon** [comment]
You can provide a top-level `effects.js` in the project root that just says `module.exports = require('./lib/effects')`. Just make sure it is included into the NPM package.

**Comment by yelouafi** [comment]
Totally forgot about module.exports. thanks!

**Comment by slorber** [comment]
>> yield getState();

> this solve the issue partly (In the real-world demo for example, the root Saga passes different selectors to sub-sagas). But I guess still better than passing getState around

> I think @slorber could have some objections here :)

I don't like the idea to query the store inside a saga but anyway many users will want to do this at some point so if it's made possible and i simply don't use it it's not a big deal. If I were to use it I would prefer something like `yield select(selector)` which hide the state shape (but both can be provided anyway). I guess the former "storeIO" now need a getState method provided too.

Not sure we are still discussing about doc here :)

**Comment by paul42** [comment]
sorry to digress from the current conversation, but as I'm reading the docs [DeclarativeEffects](http://yelouafi.github.io/redux-saga/docs/basics/DeclarativeEffects.html) what does cps stand for?

Thanks!
-Paul

**Comment by fhelwanger** [comment]
It stands for [Continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style).

**Comment by paul42** [comment]
Ah, having worked with node I recognize now - I've started with node so I've only seen them referred to as 'node-style callbacks' and it's amusing to me. It couldn't hurt, but I would say it's not necessary - I was just more curious and mistakenly assumed it was some internal designation

**Comment by ngerritsen** [comment]
Would it be nice to have some kind of block diagram? Just to show where you could use saga's in Redux.

**Comment by markerikson** [comment]
A side-by-side comparison of how you'd do something in thunks vs sagas would be helpful.  Also a description of why you would want to prefer sagas over thunks, and maybe even some bits on what generators are and how to use them.  I grok promises, and my team _mostly_ understands them, but generators I'm still pretty confused on, and sagas doubly so.

**Comment by yelouafi** [comment]
>Would it be nice to have some kind of block diagram? Just to show where you could use saga's in Redux.

That would be nice.

BTW nice article @ngerritsen. Thanks

**Comment by yelouafi** [comment]
@markerikson You may find useful [this series on Generators](https://davidwalsh.name/es6-generators) written by Kyle Simpson

**Comment by yelouafi** [comment]
FYI released 0.8.0 

https://github.com/yelouafi/redux-saga/releases/tag/v0.8.0

The release include the `takeEvery` and `takeLatest` helpers.

There are some additions on the docs

- Added a [beginner Tutorial](http://yelouafi.github.io/redux-saga/docs/introduction/BeginnerTutorial.html). No shiny demos, This is a simple tutorial for beginners I tried make the basic concepts for new users as accessible as I can.

- Rearranged the TOC. the Basics section doesnt include the `take` Effect. Event pulling is [introduced in the Advanced Section](http://yelouafi.github.io/redux-saga/docs/advanced/FutureActions.html) and then immediately followed by a section on [non blocking calls/Concurrency](http://yelouafi.github.io/redux-saga/docs/advanced/NonBlockingCalls.html) to illustrate concurrency issues.

The main reason behind this is that I think `take` shouldn't be used by new users until they become aware of concurrency issues. I'll also refactor the repo examples to use the helper functions

**Comment by yelouafi** [comment]
Forgot to mention the changes in the readme

**Comment by nickzarate** [comment]
Hey there @yelouafi I have a suggestion, what do you think about adding a section (even real short) about best practices using redux-saga? It could very well just be me (or maybe I missed it in the docs), but I am not 100% sure how to organize my sagas, for instance which to dynamically call sagas and which sagas to call at the beginning, and how to use sagas together. I read somewhere, and I can't find it again, where you mentioned a separation between worker sagas and watcher sagas, maybe elaborate a bit more on that? In terms of which to dynamically call, I'm wondering what is better, to start one saga and branch off from there? Or fire off all possible sagas at once? Or maybe a combination? Again, I might just be missing something here.

**Comment by yelouafi** [comment]
@nickzarate Maybe it's a little early for best practices as we need to get feedback from people using it in real world apps. But I agree we'll have to add some 'advices' in the recipes section.

Actually i'm thinking of a few But there maybe better opinions

Testability

- Use declarative call form `yield call(fn,...)` instead of `yield fn(...)` 
- Use named functions in `take(fn)` instead of inlined `take(action => ...)`.
- Avoid using `call(fn.bind(...))` as `bind` returns a different reference on each call  

Work Organisation
- Separation watcher/worker
- Favor high level API `takeEvery`/`takeLatest`
- Use `take` only when necessary (non trivial flow) and preferably to build other Higher level APIs

There maybe others related to how to use Store state, connect to external event emitters ...

**Comment by guybenron** [comment]
Came here looking for best practices. I'd love to see the following in the revised docs:

* Where should the line cross between reducers and sagas? I've seen two basic approaches: split responsibility according to some guideline (sync/async, UI/backend), or have the sagas handle everything and dispatch an `updateState` action for a dumb reducer to incorporate the new state into the store.

* How should independent sagas initiate each other? For example, a login saga and an API fetch saga which runs post-login. Should the login saga `put` a `fetchAPI` action? Should the fetch saga `take` a `loginSuccess` action? Should there be a third saga responsible for this?

**Comment by yelouafi** [comment]
Closing this. It was mainly to track the migration of docs from single README to gitbook. Sure there are still things to be added but I think this is an ongoing process that can't be tracked in ad-hoc issue. it's better to create specific issues for each use case using the `docs` label
